client/consensus/grandpa/justification.go

```diff
@@ -0,0 +1,313 @@
+// Copyright 2023 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"errors"
+	"fmt"
+	"reflect"
+
+	finalityGrandpa "github.com/ChainSafe/gossamer/pkg/finality-grandpa"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+	"golang.org/x/exp/constraints"
+)
+
+var (
+	errInvalidAuthoritiesSet    = errors.New("current state of blockchain has invalid authorities set")
+	errBadJustification         = errors.New("bad justification for header")
+	errBlockNotDescendentOfBase = errors.New("block not descendent of base")
+)
+
+// Justification A GRANDPA justification for block finality, it includes a commit message and
+// an ancestry proof including all headers routing all precommit target blocks
+// to the commit target block. Due to the current voting strategy the precommit
+// targets should be the same as the commit target, since honest voters don't
+// vote past authority set change blocks.
+//
+// This is meant to be stored in the db and passed around the network to other
+// nodes, and are used by syncing nodes to prove authority set handoffs.
+type Justification[
```

Comment:
Any reason why you didn't just make it name the same as substrate (`GrandpaJustification`)?

Reply:
Yeah, the editor complains that the type name begins with the package name.